### PR TITLE
Add support for CXX_STANDARD

### PR DESCRIPTION
**Pull request by chfast**

### CODE REVIEW

File: cmake/modules/hunter_dump_cmake_flags.cmake [hunk 1]
```diff
@@ -53,6 +53,35 @@ function(hunter_dump_cmake_flags)
     endforeach()
   endif()
 
+
+  # Set the C / C++ standard flag.
+  # CMAKE_<lang>_STANDARD_REQUIRED flag is ignored.
+  # This maps the logic in the CMake code:
+  # https://github.com/Kitware/CMake/blob/master/Source/cmLocalGenerator.cxx#L1433-L1467
```
Comment (ruslo) [inline cmake/modules/hunter_dump_cmake_flags.cmake]: Use stable link here, e.g.:
* https://github.com/Kitware/CMake/blob/10cbb87fa3a5f81dc9dedd24580a64e062047887/Source/cmLocalGenerator.cxx#L1433-L1467

File: cmake/projects/Qt/qt3d/hunter.cmake [hunk 2]
```diff
@@ -88,5 +88,5 @@ hunter_pick_scheme(DEFAULT url_sha1_qt)
 hunter_download(
     PACKAGE_NAME Qt
     PACKAGE_COMPONENT "qt3d"
-    PACKAGE_INTERNAL_DEPS_ID "10"
```
Comment (ruslo) [inline cmake/projects/Qt/qt3d/hunter.cmake]: For good or for bad but Qt is not using `hunter_dump_cmake_flags` so PACKAGE_INTERNAL_DEPS_ID should not be changed.

File: cmake/projects/glfw/hunter.cmake [hunk 3]
```diff
@@ -25,7 +25,7 @@ hunter_pick_scheme(DEFAULT url_sha1_cmake)
 hunter_cacheable(glfw)
 hunter_download(
     PACKAGE_NAME glfw
-    PACKAGE_INTERNAL_DEPS_ID "1"
```
Comment (ruslo) [inline cmake/projects/glfw/hunter.cmake]: This package use standard CMake scheme so PACKAGE_INTERNAL_DEPS_ID should not be changed.

**Comment by ruslo** [comment]
OpenSSL use `hunter_dump_cmake_flags`, please update `PACKAGE_INTERNAL_DEPS_ID` in `OpenSSL/hunter.cmake`.

**Comment by chfast** [comment]
Ok, thanks for reviewing this. Only these 3 issues about PACKAGE_INTERNAL_DEPS_ID? I bumped all I found :)

**Comment by ruslo** [comment]
> I bumped all I found :)

We changed `hunter_dump_cmake_flags` so we have to change all schemes that affected. It's autotools schemes, Boost and OpenSSL. Qt is not in this list. CMake-based projects too.

**Comment by chfast** [comment]
I till take care of this soon.

**Comment by chfast** [comment]
Fixed. I hope I found everything.

File: cmake/modules/hunter_get_lang_standard_flag.cmake [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  set(std "${CMAKE_${LANG}_STANDARD}")
+  if(std)
+    set(ext "EXTENSION")  # By default extensions are assumed On.
+    if(DEFINED CMAKE_${LANG}_EXTENSIONS AND NOT CMAKE_${LANG}_EXTENSIONS)
+      set(ext "STANDARD")
+    endif()
+
+    set(standards "${${LANG}_standards}")
+    list(FIND standards "${std}" begin)
+    if(NOT "${begin}" EQUAL -1)
```
Comment (ruslo) [inline cmake/modules/hunter_get_lang_standard_flag.cmake]: Same trick here:
```cmake
if("${begin}" EQUAL -1)
  hunter_internal_error("Standard not found")
endif()
```

File: cmake/modules/hunter_get_lang_standard_flag.cmake [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+      list(LENGTH standards end)
+      math(EXPR end "${end} - 1")
+      foreach(idx RANGE ${begin} ${end})
+        list(GET standards ${idx} std)
+        set(option_name "CMAKE_${LANG}${std}_${ext}_COMPILE_OPTION")
+        set(flag "${${option_name}}")
+        string(COMPARE NOTEQUAL "${flag}" "" has_flag)
+        if(has_flag OR CMAKE_${LANG}_STANDARD_REQUIRED)
```
Comment (ruslo) [inline cmake/modules/hunter_get_lang_standard_flag.cmake]: I think we should follow CMake C++ code logic:
* https://github.com/Kitware/CMake/blob/3bccdd89c88864839a0c8d4ea56bd069c90fa02b/Source/cmLocalGenerator.cxx#L1448-L1467 if STANDARD_REQUIRED is ON try the exact version. If flags found - use them, if not - fatal error.
* https://github.com/Kitware/CMake/blob/3bccdd89c88864839a0c8d4ea56bd069c90fa02b/Source/cmLocalGenerator.cxx#L1448-L1467 if the version we asking for is less or equal to default version, then use these flags (even if they are empty)
* https://github.com/Kitware/CMake/blob/3bccdd89c88864839a0c8d4ea56bd069c90fa02b/Source/cmLocalGenerator.cxx#L1448-L1467 try version lower than asking one by one until we find the non-empty flags

File: cmake/modules/hunter_get_lang_standard_flag.cmake [hunk 6]
```diff
@@ -0,0 +1,40 @@
+# Copyright (c) 2017 Pawel Bylica
+# All rights reserved.
+
+include(hunter_status_debug)
+
+function(hunter_get_lang_standard_flag LANG OUTPUT)
```
Comment (chfast) [inline cmake/modules/hunter_get_lang_standard_flag.cmake]: Yep, I was thinking about it, be we should also reset the `OUTPUT` var first?

Comment (ruslo) [inline cmake/modules/hunter_get_lang_standard_flag.cmake]: We can, I guess it doesn't hurt.

File: cmake/modules/hunter_get_lang_standard_flag.cmake (hunk 5 above)
Comment (chfast) [inline cmake/modules/hunter_get_lang_standard_flag.cmake]: I try version lower. If "17" is empty we try "14", then "11", then "98". As in https://github.com/Kitware/CMake/blob/3bccdd89c88864839a0c8d4ea56bd069c90fa02b/Source/cmLocalGenerator.cxx#L1521-L1530.

I just found a way to place everything inside a single loop. This makes the functions much smaller. CMake programming is quite explicit :)

File: cmake/modules/hunter_get_lang_standard_flag.cmake (hunk 6 above)
Comment (ruslo) [inline cmake/modules/hunter_get_lang_standard_flag.cmake]: Also I think for debugging purpose we can print next variables using `hunter_status_debug`:
* CMAKE_${LANG}_STANDARD_DEFAULT
* CMAKE_${LANG}_STANDARD
* CMAKE_${LANG}_EXTENSIONS
* CMAKE_${LANG}_STANDARD_REQUIRED

File: cmake/modules/hunter_get_lang_standard_flag.cmake (hunk 5 above)
Comment (ruslo) [inline cmake/modules/hunter_get_lang_standard_flag.cmake]: > This makes the functions much smaller

Okay, let try this version, if there will be bugs discovered we will use more pedantic approach.

File: cmake/modules/hunter_get_lang_standard_flag.cmake [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+      list(LENGTH standards end)
+      math(EXPR end "${end} - 1")
+      foreach(idx RANGE ${begin} ${end})
+        list(GET standards ${idx} std)
+        set(option_name "CMAKE_${LANG}${std}_${ext}_COMPILE_OPTION")
+        set(flag "${${option_name}}")
+        string(COMPARE NOTEQUAL "${flag}" "" has_flag)
+        if(has_flag OR CMAKE_${LANG}_STANDARD_REQUIRED)
+          # Break if flag found or standard is required and we don't want to
+          # continue checking older standards.
```
Comment (ruslo) [inline cmake/modules/hunter_get_lang_standard_flag.cmake]: Add for debug purposes:
```cmake
hunter_status_debug("${option_name}: '${flag}'")
```

**Comment by chfast** [comment]
Requested changes applied. Sorry for delay, I took holidays.

File: cmake/modules/hunter_get_lang_standard_flag.cmake [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+  hunter_status_debug("CMAKE_${LANG}_STANDARD_DEFAULT: ${CMAKE_${LANG}_STANDARD_DEFAULT}")
+  hunter_status_debug("CMAKE_${LANG}_STANDARD: ${CMAKE_${LANG}_STANDARD}")
+  hunter_status_debug("CMAKE_${LANG}_EXTENSIONS: ${CMAKE_${LANG}_EXTENSIONS}")
+  hunter_status_debug("CMAKE_${LANG}_STANDARD_REQUIRED: ${CMAKE_${LANG}_STANDARD_REQUIRED}")
+
+  set("${OUTPUT}" "" PARENT_SCOPE)  # Reset output in case of quick return.
+
+  string(COMPARE EQUAL "${CMAKE_${LANG}_STANDARD_DEFAULT}" "" no_default)
+  if(no_default)
+    # This compiler has no notion of language standard levels.
```
Comment (ruslo) [inline cmake/modules/hunter_get_lang_standard_flag.cmake]: Please add this link as a comment:
* https://github.com/Kitware/CMake/blob/3bccdd89c88864839a0c8d4ea56bd069c90fa02b/Source/cmLocalGenerator.cxx#L1427-L1432

File: cmake/modules/hunter_get_lang_standard_flag.cmake [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    return()
+  endif()
+
+  set(standard "${CMAKE_${LANG}_STANDARD}")
+  string(COMPARE EQUAL "${standard}" "" no_standard)
+  if(no_standard)
+    # The standard not defined by user.
```
Comment (ruslo) [inline cmake/modules/hunter_get_lang_standard_flag.cmake]: Please add this link as a comment:
* https://github.com/Kitware/CMake/blob/3bccdd89c88864839a0c8d4ea56bd069c90fa02b/Source/cmLocalGenerator.cxx#L1433-L1437

File: cmake/modules/hunter_get_lang_standard_flag.cmake [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    return()
+  endif()
+
+  set(flag "")
+  set(ext "EXTENSION")  # By default extensions are assumed On.
+  if(DEFINED CMAKE_${LANG}_EXTENSIONS AND NOT CMAKE_${LANG}_EXTENSIONS)
+    set(ext "STANDARD")
```
Comment (ruslo) [inline cmake/modules/hunter_get_lang_standard_flag.cmake]: Please add this link as a comment:
* https://github.com/Kitware/CMake/blob/3bccdd89c88864839a0c8d4ea56bd069c90fa02b/Source/cmLocalGenerator.cxx#L1438-L1446

**Comment by chfast** [comment]
Done.

**Comment by ruslo** [comment]
Boost:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.1527
* https://travis-ci.org/ingenue/hunter/builds/218497773

OpenSSL:
* https://travis-ci.org/ingenue/hunter/builds/218499877
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.1528

odb-boost:
* https://travis-ci.org/ingenue/hunter/builds/218501331

**Comment by ruslo** [comment]
Uploading.

Boost:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.1531
* https://travis-ci.org/ingenue/hunter/builds/218737557

OpenSSL:
* https://travis-ci.org/ingenue/hunter/builds/218738060
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.1532

odb-boost:
* https://travis-ci.org/ingenue/hunter/builds/218738319

**Comment by ruslo** [comment]
Released: https://github.com/ruslo/hunter/releases/tag/v0.18.37

**Pull request merged**